### DISCUSSION TITLE
Should the SCA be a builtin-actor or a user-defined actor?

### DISCUSSION THREAD

**Discussion by adlrocha** [consensus-shipyard/consensuslab#155] [General]
In the MVP implementation of hierarchical consensus (HC), the SCA is a `builtin-actor` that implements the core logic of the protocol. One of the things the SCA is responsible for is the execution of cross-net messages. The reasons why we decided to make the SCA a `builtin-actor` (apart from the fact that when we started working on HC there was no FVM and we had to implement everything as `spec-actors` over the `LegacyVM`) is:
1. So the SCA is a singleton actor spawned in genesis in every network.
2. So the SCA has a specific unique address known by all network participants (currently `f064`) like the rest of builtin-actors.
3. So that cross-net messages can be executed by implicitly applying message in the SCA.

After skimming through the protocol spec, @ZenGround0 (and others) have been wondering if we could make the SCA a user-defined actor (it can simplify many things). If we remove the need for 3. to execute cross-net messages (as proposed [here](https://github.com/filecoin-project/eudico/issues/229)), and we force for the SCA to be a singleton with a well-known and specific address we may be able to have SCA as a user-defined actor (although from my I don't know the best way to do this over the FVM for user-defined actors).

The goal of this discussion is to raise the various options and agree on the best way going forward. According to the outcome we may require an update to the spec and a significant update to the implementation.

**Comment by anorth** [comment]
While I don't fully understand everything involved in the proposals here, I don't yet see any strong reason why a subnet coordinator actor should be a built-in actor in the main Filecoin chain.

- We don't need to spawn at genesis or otherwise immaculately conceive it, it can just be deployed to FVM
- It will have an address that can be easily published and known anyway
- I don't see a good reason to attempt to enforce singletonhood. It's impossible to prevent someone deploying the same code again anyway.
- I don't fully understand the implicit message application yet, but (a) that would seem to be a super major scary increase in complexity to Filecoin L1 and (b) it's not clear what property that would have that you can't get by normal message application (especially after [account abstraction](https://github.com/filecoin-project/FIPs/discussions/388)), except (c) it introduces a problem of who pays gas.

For both implicit message application and a new built-in actor, I would expect that we (stewards of FIlecoin) would want to see a robust argument about why you can't achieve similar outcomes without adding complexity to Filecoin's consensus rules (which includes built-in actors), and if such argument was robust we'd put some real thought into what minimal facility we could add that would resolve whatever was blocking it (e.g. a new VM syscall).

**Reply by adlrocha** [reply]
Thank you for the feedback @anorth. I agree with all your points. All of these were historical design decision due to _"what was there"_ when we kicked-off the project, and many no longer apply. I think we are all converging to the same proposal of making the SCA a user-defined actor with a specific address known by all HC participants (and this will definitely the case for mainnet, in subnets we have a bit more of freedom on how to deploy and address this contract as they are spawned from scratch, but this won't impact mainnet).

The main thing where we'll need some additional input is in the application of cross-net message. In the current approach, for the execution of cross-net messages we need to call the [`ApplyMsg`](https://github.com/adlrocha/builtin-actors/blob/9ede48182f541b565a63a294f3fc67a423128479/actors/hierarchical_sca/src/lib.rs#L536) function in the SCA. This function performs a set of checks according to the type of cross-net message, updates the state of the SCA, and forwards the corresponding messages to other actors. To prevent external calls to this function, and to ensure that this function can only be called when a block is validated to execute the cross-message, we force the caller to be the `SYSTEM_ACTOR`, and thus the implicit application of the message. Skimming through your _account abstraction_ proposal, I feel this could replace the `SYSTEM_ACTOR` dance that we are currently doing. IIUC we would just need to direct cross-net messages to the SCA through a gateway actor for account abstraction.

Anyway, any idea or suggestion on the best way to handle the application of this kind of messages is more thank welcome.

**Reply by anorth** [reply]
Re historical design decisions, 👍 I totally understand.

There's lots I don't understand at present about cross-net messages. E.g. it's not obvious to me why you need to prevent external calls to a function (or exactly what external means here) or what it means to ensure it's only called when a block (in which chain?) is validated. How (and why) does your design differ from the many other L2s and bridges between them in, say, the Ethereum ecosystem?

This thread may not be the right place to explain it to me. I'll read those parts of the spec again but perhaps you want to open a different issue?

**Reply by aakoshh** [reply]
I agree that we should discuss this perhaps in person, there seems to be too much jargon that mean different things to different people. For what it's worth, my current understanding is the following.

By "external call" I _think_ we mean calling the actor from another user-defined actor by passing a message, but we have some special messages that should strictly come from the consensus protocol, not from random users. It's like authorization: users are allowed to call _some_ methods on the SCA, but not all, some are privileged, and this is what so far has been reserved to the `SYSTEM_ACTOR`. We just think we will no longer be able to use `SYSTEM_ACTOR` as a `caller`, so we're looking for a different way to differentiate. 

The reason we want to prevent calls out of the blue is because of the semantics of cross-net messages. There are three kinds AFAIU: 
1) The user wants to send a message to someone in a different subnet, which is fine, they should be allowed to call that method; these are the originators of the next two types.
2) Top-down messages come from the parent subnet, which is implicitly trusted because the child subnet operates a full node on the parent subnet as well. These messages might travel through the subnet to deeper levels, or might be applied locally. But they _must_ only be applied once the block they are in is stable on the parent subnet, and they _must not_ come from a random user in the current subnet; they should be part of the parent consensus. Top down messages can increase the circulating supply of the subnet.
3) Bottom-up messages come from one of the child subnets, to the parent. The parent does not follow the children; instead a child subnet _must_ submit these messages as part of a _checkpoint_ that a Subnet Actor (SA) in the parent subnet can verify, and forward to the SCA. So again these messages _must not_ come from a random user in the current subnet. Bottom-up messages can release circulating supply back to the parent.

Indeed this is recursive generalisation of sidechains and bridges, with inherent trust in one way (child trusts the parent because of full validation) and partial trust the other way (parent trusts child checkpoints but requires collateral and checkpoint). Maybe one difference is that both sides have their own consensus protocols: the child understand the parent; but the parent needn't understand the child, only a user defined actor understands their checkpoints.

A difference between this and for example IBC is the firewall property. In Cosmos anyone can open channels to any other chain, and the quorum certificates are verified, but there's no way to slash the other chain if it misbehaves. Here the damage is limited to how much collateral the subnet has put up.

**Reply by adlrocha** [reply]
I think @aakoshh reply covers much of what I wanted to say :) In any case, @anorth, I added the following placeholder discussion to open the discussion about differences and similarities of HC with other L2 solutions (I feel it can be a useful exercise): https://github.com/protocol/ConsensusLab/discussions/156

Also, feel free to add any other topical discussion you think is worth debating. Any input you can give us is more than welcome. Thanks!
